libevm/triedb/firewood/firewood.go

```diff
@@ -0,0 +1,518 @@
+	}
+	return db
+}
+
+// New creates a new Firewood database with the given disk database and configuration.
+// Any error during creation will cause the program to exit.
+func New(config Config) (*Database, error) {
+	if err := validatePath(config.ChainDir); err != nil {
+		return nil, err
+	}
+
+	// Start the logs prior to opening the database
+	logPath := filepath.Join(config.ChainDir, logFileName)
+	if err := ffi.StartLogs(&ffi.LogConfig{Path: logPath}); err != nil {
+		// This shouldn't be a fatal error, as this can only be called once per thread.
+		// Specifically, this will return an error in unit tests.
+		log.Warn("firewood: error starting logs", "error", err)
+	}
+
+	dbPath := filepath.Join(config.ChainDir, dbFileName)
+	fw, err := ffi.New(dbPath, &ffi.Config{
+		NodeCacheEntries:     uint(config.CleanCacheSize) / 256, // TODO: estimate 256 bytes per node
+		FreeListCacheEntries: config.FreeListCacheEntries,
+		Revisions:            config.Revisions,
+		ReadCacheStrategy:    config.ReadCacheStrategy,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	currentRoot, err := fw.Root()
+	if err != nil {
+		return nil, err
+	}
+
+	return &Database{
+		Firewood:    fw,
+		proposalMap: make(map[common.Hash][]*proposal),
+		proposalTree: &proposal{
+			info: &blockInfo{
+				root: common.Hash(currentRoot),
+				hashes: map[common.Hash]struct{}{
+					{}: {}, // genesis block has empty hash
+				},
+			},
+		},
+	}, nil
+}
+
+func validatePath(dir string) error {
+	if dir == "" {
+		return errors.New("chain data directory must be set")
+	}
+
+	// Check that the directory exists
+	switch info, err := os.Stat(dir); {
+	case os.IsNotExist(err):
+		log.Info("Database directory not found, creating", "path", dir)
+		if err := os.MkdirAll(dir, 0o755); err != nil {
+			return fmt.Errorf("error creating database directory: %w", err)
+		}
+		return nil
+	case err != nil:
+		return fmt.Errorf("error checking database directory: %w", err)
+	case !info.IsDir():
+		return fmt.Errorf("database directory path is not a directory: %s", dir)
+	}
+
+	return nil
+}
+
+// Scheme returns the scheme of the database.
+// This is only used in some API calls
+// and in StateDB to avoid iterating through deleted storage tries.
+// WARNING: If anything from go-ethereum uses this,
+// it must be overwritten to use something like:
+// `_, ok := db.(*Database); if !ok { return "" }`
+// to recognize the Firewood database.
+func (*Database) Scheme() string {
+	return rawdb.HashScheme
+}
+
+// Initialized checks whether a non-empty genesis block has been written.
+func (db *Database) Initialized(common.Hash) bool {
+	rootBytes, err := db.Firewood.Root()
+	if err != nil {
+		log.Error("firewood: error getting current root", "error", err)
+		return false
+	}
+	root := common.BytesToHash(rootBytes)
+	// If the current root isn't empty, then unless the database is empty, we have a genesis block recorded.
+	return root != types.EmptyRootHash
```

Comment:
```suggestion
	root, err := db.Firewood.Root()
	if err != nil {
		log.Error("firewood: error getting current root", "error", err)
		return false
	}
	// If the current root isn't empty, then unless the database is empty, we have a genesis block recorded.
	return !bytes.Equal(root, types.EmptyRootHash[:])
```

The `common` package's conversion functions are foot-guns that have caused very subtle bugs for me in the past, but see above re returning a `[32]byte` from `Root()`.

Reply:
I fixed it, but I also made a PR in Firewood: ava-labs/firewood#1449.